index.js

```diff
@@ -382,13 +389,32 @@ Sharp.prototype.boolean = function(operand, operator) {
   } else {
     throw new Error('Invalid boolean operation ' + operator);
   }
+  if (isObject(options)) {
+    if (isDefined(options.raw)) {
+      if (
+        isObject(options.raw) &&
+        isInteger(options.raw.width) && inRange(options.raw.width, 1, maximum.width) &&
+        isInteger(options.raw.height) && inRange(options.raw.height, 1, maximum.height) &&
+        isInteger(options.raw.channels) && inRange(options.raw.channels, 1, 4)
+      ) {
+        this.options.booleanRawWidth = options.raw.width;
+        this.options.booleanRawHeight = options.raw.height;
+        this.options.booleanRawChannels = options.raw.channels;
+      } else {
+        throw new Error('Boolean: expected width, height and channels for raw pixel input');
+      }
+    }
+  }
+
   return this;
 };
 
 /*
   Overlay with another image, using an optional gravity
 */
 Sharp.prototype.overlayWith = function(overlay, options) {
+  /* jshint maxcomplexity:16 */
```

Comment:
We'll need much higher test coverage at this complexity level. However a better approach here is probably to abstract the common logic for the various file/Buffer/raw input to a separate function and keep the `boolean` and `overlayWith` functions themselves simpler.